### ISSUE TITLE
AsSuperVisitor: type is not an erased subtype of supertype with wildcards

### ISSUE THREAD

**Issue by wmdietlGC** [typetools/checker-framework#1111]
Take:

```
import java.util.List;

class WildcardCrash {
  void foo(Box<? super Integer> box, List<Integer> list) {
    bar(box, list);
  }

  <T extends Number> void bar(
      Box<T> box, Iterable<? extends T> list) {}

  class Box<T extends Number> {}
}
```

and run any checker (tested with regex and nullness) and you'll get:

```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: Number
  superType: Integer
  Compilation unit: WildcardCrash.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:725)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:41)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:155)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:340)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:29)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:86)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:68)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:131)
  org.checkerframework.framework.type.DefaultTypeHierarchy.castedAsSuper(DefaultTypeHierarchy.java:1063)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:493)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:52)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:257)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcardSupertype(DefaultTypeHierarchy.java:998)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Wildcard(DefaultTypeHierarchy.java:622)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Wildcard(DefaultTypeHierarchy.java:52)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:346)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:257)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcardSubtype(DefaultTypeHierarchy.java:1037)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcard_Wildcard(DefaultTypeHierarchy.java:946)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitWildcard_Wildcard(DefaultTypeHierarchy.java:52)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:716)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:257)
  org.checkerframework.framework.type.DefaultTypeHierarchy.checkAndSubtype(DefaultTypeHierarchy.java:349)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isContainedBy(DefaultTypeHierarchy.java:432)
  org.checkerframework.framework.type.DefaultTypeHierarchy.compareTypeArgs(DefaultTypeHierarchy.java:569)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitTypeArgs(DefaultTypeHierarchy.java:541)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:505)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:52)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:257)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:192)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:170)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:1667)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:1634)
  org.checkerframework.common.basetype.BaseTypeVisitor.checkArguments(BaseTypeVisitor.java:2080)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethodInvocation(BaseTypeVisitor.java:819)
  org.checkerframework.checker.regex.RegexVisitor.visitMethodInvocation(RegexVisitor.java:110)
  org.checkerframework.checker.regex.RegexVisitor.visitMethodInvocation(RegexVisitor.java:36)
```

**Comment by entonio** [comment]
Hi,

The following code:
```java
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;

@Configuration
public class SpringSecurityConfig extends WebSecurityConfigurerAdapter {

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http
            .csrf()
            .disable()
            .authorizeRequests()
            .antMatchers("/g/**")
            .fullyAuthenticated().and().httpBasic();
    }
}

```

results in:
```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: HttpSecurityBuilder<H extends HttpSecurityBuilder<H>>
  superType: HttpSecurity; invoke the compiler with -AprintErrorStack to see the stack trace.
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @Initialized @NonNull HttpSecurityBuilder<H extends @Initialized @NonNull HttpSecurityBuilder<H>>
  superType: @Initialized @NonNull HttpSecurity; invoke the compiler with -AprintErrorStack to see the stack trace.
2 errors
```

Now, I can't guarantee the code is 100% correct. But maybe it is, even. It certainly works as intended, tho of course that's a separate issue. But my issue is with CF outputting an **error** instead of a warning. I've got `-Awarns` set. If javac is able to build the project, then CF shouldn't be preventing it.

**Comment by smillst** [comment]
@entonio Sorry you're having problems.  I can't reproduce this error because the test case is not complete.  Please follow the instructions in [the manual](https://checkerframework.org/manual/#reporting-bugs) and, as this is a closed issue,  please create a new issue with a complete test case.  Thanks!

**Comment by entonio** [comment]
My intention was not so much to solve this specific issue, but to point out that when CF 'crashes' it shouldn't break compilation, at least when `-Awarns` is set. From what I understand, you sometimes try/catch these issues, I guess I was arguing for a global catch, if such a thing is possible within the architecture.

**Comment by wmdietlGC** [comment]
We have a global try/catch, but then raise an error: https://github.com/typetools/checker-framework/blob/master/framework/src/org/checkerframework/framework/source/SourceChecker.java#L794
My attitude is that these exceptions happen very rarely and we're working very hard to fix them as soon as possible after a report.
If we changed these to warnings, people might start ignoring them instead of reporting them.
But if others feel differently, it would be easy to check for `-Awarns` in that location.

**Comment by entonio** [comment]
> My attitude is that these exceptions happen very rarely and we're working very hard to fix them as soon as possible after a report.
If we changed these to warnings, people might start ignoring them instead of reporting them.

Oh, ok, I can totally empathise with that.
If the next version  doesn't fix the issue I'll try to submit a test case.

**Comment by entonio** [comment]
(In this specific case, the issue happened with a colleague, he spent a fair amount of time to find out that it was down to CF, and then the solution was to disable CF. I've been trying to push CF whenever I can, but when stuff like this happens a large amount of goodwill goes out of the window 😳)

**Comment by mernst** [comment]
I'm sorry to hear that it took a while to figure out that the problem was the CF.
Is there anything that we could do to help speed up such a diagnosis?  It's easy for me to tell from the error message and stack trace, but obviously it's more confusing to others.
Thanks for any suggestions!

**Comment by entonio** [comment]
The output is as follows:

```
Executing: gradle clean build

:clean
:generateBuildInfo
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: HttpSecurityBuilder<H extends HttpSecurityBuilder<H>>
  superType: HttpSecurity; invoke the compiler with -AprintErrorStack to see the stack trace.
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @Initialized @NonNull HttpSecurityBuilder<H extends @Initialized @NonNull HttpSecurityBuilder<H>>
  superType: @Initialized @NonNull HttpSecurity; invoke the compiler with -AprintErrorStack to see the stack trace.
2 errors
:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --info or --debug option to get more log output.


BUILD FAILED

Total time: 5.781 secs


Build failure (see the Notifications window for stacktrace): gradle clean build
```

If it were me, the `-A` suggestions and the `@Initialized @NonNull` would have betrayed that it was a CF output. The error when using `-Awarn` would have suggested that it wasn't javac refusing to compile. But my colleague, not having that much experience with annotation processors, spent a lot of time instead trying to debug the code itself, which is complicated because it's a lot of generics. Some message saying

```
Checker Framework crashed - this is not a javac error, it may be possible to build by disabling Checker Framework. We however suggest you report this issue so we can work on it.
```

would have been more than enough. In fact, if he did end up not noticing it and losing some time before he did, he'd then say 'd'oh, they're trying to help and I didn't even notice' and would have been impressed. I think.

**Comment by mernst** [comment]
@entonio , thanks for the concrete suggestion.
@smillst , could you implement the suggestion when you get a chance?

**Comment by jhovell** [comment]
Why is this issue closed? Is this bug fixed in a more recent version of CF? Or should requesting a fix for this crash be a separate issue?

**Comment by jhovell** [comment]
Linking to this issue since it seems related. https://github.com/typetools/checker-framework/issues/1312

**Comment by wmdietl** [comment]
The original issue was fixed by #1115 and the crash message was improved by #1422.
If you're running into a crash, please file a new issue with a test case.
